networkx/algorithms/distance_measures.py

```diff
@@ -300,6 +302,65 @@ def diameter(G, e=None, usebounds=False):
     return max(e.values())
 
 
+def harmonic_diameter(G, sp=None):
+    """Returns the harmonic diameter of the graph G.
+
+    The harmonic diameter of a graph is the harmonic mean of the distances
+    between all pairs of distinct vertices. Graphs that are not strongly
+    connected have infinite diameter and mean distance, making such
+    measures not useful. Restricting the diameter or mean distance to
+    finite distances yields paradoxical values (e.g., a perfect match
+    would have diameter one). The harmonic mean handles gracefully
+    infinite distances (e.g., a perfect match has harmonic diameter equal
+    to the number of vertices minus one), making it possible to assign a
+    meaningful value to all graphs.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+       A graph
+
+    sp : dict of dicts, optional
+       All-pairs shortest path lengths as a dictionary of dictionaries
+
+    Returns
+    -------
+    hd : float
+       Harmonic diameter of graph
+
+    References
+    ----------
+    .. [1] Massimo Marchiori and Vito Latora, "Harmony in the small-world".
+           *Physica A: Statistical Mechanics and Its Applications*
+           285(3-4), pages 539-546, 2000.
+           <https://doi.org/10.1016/S0378-4371(00)00311-3>
+    """
+    order = G.order()
+
+    sum_invd = 0
+    for n in G:
+        if sp is None:
+            length = nx.single_source_shortest_path_length(G, n)
```

Comment:
This would not work for `weighted` graphs, is `harmonic_diameter` defined for a weighted graph?

Reply:
Ah OK I see I've got confused between `single_source_shortest_path_length` and `shortest_path_length`. The implementation as such is correct for distances defined as shortest paths. You can use another notion of distance—nothing prevents you from taking the harmonic mean. I thought that NetworkX would automatically "understand" whether a graph was weighted and compute distances accordingly (as `diameter` and `eccentricity` are already implemented in that way), but now I'm seeing that, say, `harmonic_centrality` has an explicitly `distance` parameter.

It seems that different people implemented different APIs in NetworkX for using a non-standard metric: it is passed via a dictionary of distances in `diameter`, and via a property used to detect edge weights in `harmonic_centrality`.

I'd stick with the current implementation as it mimics that of `diameter`, but you guys know better than me.

Reply:
It looks like this module (`distance_measures.py`) could use a pass through to update all functions to allow weighted graphs. The definition should be straightforward, following from the definition in shortest_path. 
I'll open an issue for this suggestion separate from this PR.